File: 백준/Gold/12784. 인하니카 공화국/인하니카 공화국.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static int[] dp;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {

            StringTokenizer st = new StringTokenizer(br.readLine());

            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());

            List<Node>[] tree = new ArrayList[n];
            dp = new int[n];

            for (int i = 0; i < n; i++) {
                tree[i] = new ArrayList<>();
            }

            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());

                int u = Integer.parseInt(st.nextToken()) - 1;
                int v = Integer.parseInt(st.nextToken()) - 1;
                int d = Integer.parseInt(st.nextToken());

                tree[u].add(new Node(v, d));
                tree[v].add(new Node(u, d));
            }
            
            sb.append(dfs(0, -1, tree)).append("\n");
        }

        System.out.print(sb);
    }

    private static int dfs(int node, int parent, List<Node>[] tree) {

        if (tree[node].size() == 1 && tree[node].get(0).node == parent) {
            return dp[node] = tree[node].get(0).dynamite;
        }

        for (Node child : tree[node]) {
            int nextNode = child.node;

            int cost = 0;
            if (nextNode != parent) {
                cost = dfs(nextNode, node, tree);
            }

            dp[node] += Math.min(cost, child.dynamite);
        }

        return dp[node];
    }

    static class Node {

        int node, dynamite;

        public Node(int node, int dynamite) {
            this.node = node;
            this.dynamite = dynamite;
        }
    }
}

File: 백준/Gold/12784. 인하니카 공화국/인하니카 공화국.py
import sys
from collections import defaultdict

input = sys.stdin.readline
dp = []


def dfs(node, parent, tree):
    if len(tree[node]) == 1 and tree[node][0][0] == parent:
        dp[node] = tree[node][0][1]
        return dp[node]

    for next_node, dynamite in tree[node]:

        if next_node != parent:
            cost = dfs(next_node, node, tree)
            dp[node] += min(cost, dynamite)

    return dp[node]


t = int(input())

ans = []

for _ in range(t):

    n, m = map(int, input().split())

    tree = defaultdict(list)
    dp = [0] * n

    for _ in range(m):
        u, v, d = map(int, input().split())
        u -= 1
        v -= 1

        tree[u].append((v, d))
        tree[v].append((u, d))

    ans.append(dfs(0, -1, tree))

print("\n".join(map(str, ans)))
